### PR TITLE
Remove default AffinityKeyName values

### LINKED ISSUE #976: Affinity conflicts between clusters
The affinity flow looks like this:
request -> route -> cluster -> check affinity -> load balance -> proxy request -> establish affinity to destination -> proxy response

However, this can run into conflicts when you have multiple clusters.
request1 -> route1 -> cluster1 -> check affinity -> load balance -> proxy request -> establish affinity to destination1 -> proxy response
request2 -> route2 -> cluster2 -> check affinity [destination1 does not exist on this cluster] -> load balance -> proxy request -> establish affinity to destination2 [overrides affinity cookie] -> proxy response

And what's worse, `CookieSessionAffinityProviderOptions` is global, you can't even specify different cookie names per route.

### PR DESCRIPTION
**Pull request by alnikola**
Removes default `AffinityKeyName` values which are customer affinity header name and affinity cookie name. Now, the user must always explicitly set `SessionAffinityConfig.AffinityKeyName` property.

Fixes #976

### CODE REVIEW

File: src/ReverseProxy/Service/SessionAffinity/CustomHeaderSessionAffinityProvider.cs [hunk 1]
```diff
@@ -30,7 +28,7 @@ protected override string GetDestinationAffinityKey(DestinationState destination
 
         protected override (string? Key, bool ExtractedSuccessfully) GetRequestAffinityKey(HttpContext context, SessionAffinityConfig config)
         {
-            var customHeaderName = config.AffinityKeyName ?? DefaultCustomHeaderName;
+            var customHeaderName = config.AffinityKeyName!;
```
Comment (MihaZupan) [inline src/ReverseProxy/Service/SessionAffinity/CustomHeaderSessionAffinityProvider.cs:31]: I assume we want to make `AffinityKeyName` on `SessionAffinityConfig` non-nullable now? That would also avoid using `!` everywhere.

Comment (alnikola) [reply src/ReverseProxy/Service/SessionAffinity/CustomHeaderSessionAffinityProvider.cs:31]: I tried this, but it didn't work due to a delayed initialization. Let me check it again, may be I manage to do it.

Comment (alnikola) [reply src/ReverseProxy/Service/SessionAffinity/CustomHeaderSessionAffinityProvider.cs]: OK. I did it.

Comment (alnikola) [reply src/ReverseProxy/Service/SessionAffinity/CustomHeaderSessionAffinityProvider.cs]: Hmm, there is issue with JSON serialization. It fails with message

>"Deserialization of reference types without parameterless constructor is not supported."

Comment (MihaZupan) [reply src/ReverseProxy/Service/SessionAffinity/CustomHeaderSessionAffinityProvider.cs]: We don't have ctors on other records with required properties - only using the `!` operator on the type itself

Comment (Tratcher) [reply src/ReverseProxy/Service/SessionAffinity/CustomHeaderSessionAffinityProvider.cs]: The pattern we've used elsewhere for required paramters is like this:
https://github.com/microsoft/reverse-proxy/blob/d6d8374664bd86b5be9a75be212bf9911df4292f/src/ReverseProxy/Abstractions/DestinationConfig.cs#L19

And then validating its presence in the config validator.

Adding constructors breaks serialization.

File: src/ReverseProxy/Abstractions/SessionAffinityConfig.cs [hunk 2]
```diff
@@ -10,6 +10,12 @@ namespace Yarp.ReverseProxy.Abstractions
     /// </summary>
     public sealed record SessionAffinityConfig
     {
+
+        public SessionAffinityConfig(string affinityKeyName)
+        {
+            AffinityKeyName = affinityKeyName;
+        }
+
```
Comment (Tratcher) [inline src/ReverseProxy/Abstractions/SessionAffinityConfig.cs:18]: Remove, this breaks seralization.

File: src/ReverseProxy/Abstractions/SessionAffinityConfig.cs [hunk 3]
```diff
@@ -33,7 +39,7 @@ public sealed record SessionAffinityConfig
         /// This value should be unique across clusters to avoid affinity conflicts.
         /// https://github.com/microsoft/reverse-proxy/issues/976
         /// </summary>
-        public string? AffinityKeyName { get; init; }
+        public string AffinityKeyName { get; init; }
```
Comment (Tratcher) [inline src/ReverseProxy/Abstractions/SessionAffinityConfig.cs:42]: Update the comment to indicate this is required.
```suggestion
        public string AffinityKeyName { get; init; } = default!;
```

File: src/ReverseProxy/Service/SessionAffinity/CustomHeaderSessionAffinityProvider.cs (hunk 1 above)
Comment (alnikola) [reply src/ReverseProxy/Service/SessionAffinity/CustomHeaderSessionAffinityProvider.cs]: Done.

File: src/ReverseProxy/Abstractions/SessionAffinityConfig.cs (hunk 3 above)
Comment (alnikola) [reply src/ReverseProxy/Abstractions/SessionAffinityConfig.cs]: Done

File: src/ReverseProxy/Abstractions/SessionAffinityConfig.cs [hunk 4]
```diff
@@ -32,8 +33,9 @@ public sealed record SessionAffinityConfig
         /// The provider will give its own default if no value is set.
         /// This value should be unique across clusters to avoid affinity conflicts.
         /// https://github.com/microsoft/reverse-proxy/issues/976
+        /// This property must be set right after creating a new <see cref="SessionAffinityConfig"/> instance.
         /// </summary>
-        public string? AffinityKeyName { get; init; }
+        public string AffinityKeyName { get; init; } = default!;
```
Comment (MihaZupan) [inline src/ReverseProxy/Abstractions/SessionAffinityConfig.cs:38]: Nit: In other places we've just said "This field is required. "

https://github.com/microsoft/reverse-proxy/blob/d6d8374664bd86b5be9a75be212bf9911df4292f/src/ReverseProxy/Abstractions/ClusterConfig.cs#L16-L20

File: src/ReverseProxy/Service/Config/ConfigValidator.cs [hunk 5]
```diff
@@ -329,6 +329,11 @@ private void ValidateSessionAffinity(IList<Exception> errors, ClusterConfig clus
                 errors.Add(new ArgumentException($"No matching IAffinityFailurePolicy found for the affinity failure policy name '{affinityFailurePolicy}' set on the cluster '{cluster.ClusterId}'."));
             }
 
+            if (string.IsNullOrEmpty(cluster.SessionAffinity.AffinityKeyName))
+            {
+                errors.Add(new ArgumentException($"Affinity key name set on the cluster '{cluster.ClusterId}' must not be null."));
+            }
```
Comment (MihaZupan) [inline src/ReverseProxy/Service/Config/ConfigValidator.cs:335]: Not part of this PR, but this is another place where we could be doing more validation (making sure the Key is valid for use as header/cookie name).

Semi-related: #797 (early validation)

**Review by MihaZupan — APPROVED** [review]

File: src/ReverseProxy/Abstractions/SessionAffinityConfig.cs (hunk 4 above)
Comment (alnikola) [reply src/ReverseProxy/Abstractions/SessionAffinityConfig.cs]: Fixed
